### DISCUSSION TITLE
Slice of Struct or Slice of Pointers

### DISCUSSION THREAD

**Discussion by zibi94** [aarondl/sqlboiler#1345] [General]
Why was it decided to return a slice of pointers instead of a slice of structures.
https://github.com/volatiletech/sqlboiler/blob/35bb36b0b678dec3b360183ec76d8eb6467ab60c/templates/main/01_types.go.tpl#L19

Referring to the articles:
https://andrii-kushch.medium.com/go-what-to-return-a-slice-of-structs-vs-a-slice-of-pointers-42647912530a
https://medium.com/@philpearl/bad-go-slices-of-pointers-ed3c06b8bb41

The authors' conclusion is that it is better to use a slice of structures.

I would like to ask if there was any particular reason for this decision.

Maybe this is an idea for improvement.

**Comment by stephenafamo** [comment]
When I started maintaining SQLBoiler, things were already this way and I've never thought about it.

While I would like to dig into this and change it if there are significant benefits, it would be a **huge** breaking change so it cannot even be considered before we are ready for a new major version.

**Comment by adatob** [comment]
From the `queries/reflect.go` :
```
// Valid types to bind to are: *Struct, []*Struct, and []Struct. Keep in mind
// if you use []Struct that Bind will be doing copy-by-value as a method
// of keeping heap memory usage low which means if your Struct contains
// reference types/pointers you will see incorrect results, do not use
// []Struct with a Struct with reference types.
```

I also noticed incorrect results after binding to `[]Struct` when I used the custom reference type field inside for `jsonb` type column scanning. Using the `[]*Struct` the results are correct.

I believe this is part of the reason for using `[]*Struct` in models.

**Comment by xrn** [comment]
@stephenafamo how are you looking into this at https://github.com/stephenafamo/bob project?

**Reply by stephenafamo** [reply]
Bob also uses a slice of pointers. This is for two reasons:

1. **Bob** heavily uses generics, and it is very difficult to express getting a slice to the non-pointer version of another type. It is possible, but would make things very difficult to read, and by extension, debug.
2. Since the "one" methods all act on the pointer, having a slice of structs may be annoying to use when you have the slice and need to use a singular struct.

**Reply by xrn** [reply]
Thank you for explanation
